### PR TITLE
doc: update cc_ssh clarify host and auth keys

### PR DESCRIPTION
**Pull request by powersj**
LP: #1827021

### CODE REVIEW

File: cloudinit/config/cc_ssh.py [hunk 1]
```diff
@@ -9,11 +9,13 @@
 """
 SSH
 ---
-**Summary:** configure ssh and ssh keys
+**Summary:** configure ssh and ssh keys (host and authorized)
 
-This module handles most configuration for ssh and ssh keys. Many images have
-default ssh keys, which can be removed using ``ssh_deletekeys``. Since removing
-default keys is usually the desired behavior this option is enabled by default.
+This module handles most configuration for ssh and both host and authroized ssh
```
Comment (nilsmeyer) [inline cloudinit/config/cc_ssh.py]: typo: authorized

Comment (powersj) [reply cloudinit/config/cc_ssh.py]: fixed, thanks for the review

File: cloudinit/config/cc_ssh.py [hunk 2]
```diff
@@ -9,11 +9,13 @@
...
 SSH
 ---
-**Summary:** configure ssh and ssh keys
+**Summary:** configure ssh and ssh keys (host and authorized)
 
-This module handles most configuration for ssh and ssh keys. Many images have
-default ssh keys, which can be removed using ``ssh_deletekeys``. Since removing
-default keys is usually the desired behavior this option is enabled by default.
+This module handles most configuration for ssh and both host and authorized ssh
+keys. Many images have default ssh keys, which can be removed using
```
Comment (OddBloke) [inline cloudinit/config/cc_ssh.py]:
```suggestion
keys. Many images have default host ssh keys, which can be removed using
```

Looking at it, perhaps "Many images" should be the start of a new paragraph, as it doesn't apply to authorized keys at all.

File: cloudinit/config/cc_ssh.py [hunk 3]
```diff
@@ -9,11 +9,13 @@
...
+``ssh_deletekeys``. This prevents re-use of a private key from an image on
```
Comment (OddBloke) [inline cloudinit/config/cc_ssh.py]:
```suggestion
``ssh_deletekeys``. This prevents re-use of a private host key from an image on
```

File: cloudinit/config/cc_ssh.py [hunk 4]
```diff
@@ -9,11 +9,13 @@
...
+multiple machines. Since removing default keys is usually the desired behavior
```
Comment (OddBloke) [inline cloudinit/config/cc_ssh.py]:
```suggestion
multiple machines. Since removing default host keys is usually the desired behavior
```

**Comment by powersj** [comment]
Ok I've updated the repo based on @OddBloke's comments and added definitions for both authorized and host keys. Honestly adding those definitions helped me too. Most of the diff is due to changes in line length.

File: cloudinit/config/cc_ssh.py [hunk 5]
```diff
@@ -9,19 +9,26 @@
...
+keys.
+
+Host keys are for authenticating a specific instance. Many images have default
+host ssh keys, which can be removed using ``ssh_deletekeys``. This prevents
+re-use of a private host key from an image on multiple machines. Since
+removing default host keys is usually the desired behavior this option is
+enabled by default.
+
+Authorized keys are a list of public SSH key pairs that are allowed to
+connect to a system.  Keys can be added using the ``ssh_keys`` configuration
```
Comment (OddBloke) [inline cloudinit/config/cc_ssh.py]: `ssh_keys` is for host keys, `ssh_authorized_keys` is for authorized keys.

Comment (powersj) [reply cloudinit/config/cc_ssh.py]: Doh. well.... maybe I didn't clear it up for myself as much as I thought I had.

**Comment by powersj** [comment]
I gave the two sections headers and moved the authorized section up as I felt it would be more relevant to users looking for docs.

File: cloudinit/config/cc_ssh.py [hunk 6]
```diff
@@ -9,43 +9,26 @@
...
+^^^^^^^^^^^^^^^
 
-.. note::
-    when specifying private keys in cloud-config, care should be taken to
-    ensure that the communication between the data source and the instance is
-    secure
-
-.. note::
-    to specify multiline private keys, use yaml multiline syntax
+Authorized keys are a list of public SSH key pairs that are allowed to
```
Comment (OddBloke) [inline cloudinit/config/cc_ssh.py]:
```suggestion
Authorized keys are a list of public SSH keys that are allowed to
```

A key pair is the public key and the private key.

File: cloudinit/config/cc_ssh.py [hunk 7]
```diff
@@ -9,43 +9,26 @@
...
+connect to a system. Authorized keys for the default user/first user defined
```
Comment (OddBloke) [inline cloudinit/config/cc_ssh.py]: I think "first user defined in ``users``" is a touch long for a /. Perhaps a parenthetical:

```
for the default user (or the first user defined in ``users``) can be specified
```

File: cloudinit/config/cc_ssh.py [hunk 8]
```diff
@@ -9,43 +9,26 @@
...
-``ssh-keygen``. By default one public/private pair of each supported key type
-will be generated. The key types to generate can be specified using the
-``ssh_genkeytypes`` config flag, which accepts a list of key types to use. For
-each key type for which this module has been instructed to create a keypair, if
-a key of the same type is already present on the system (i.e. if
-``ssh_deletekeys`` was false), no key will be generated.
-
-Supported key types for the ``ssh_keys`` and the ``ssh_genkeytypes`` config
-flags are:
+Importing ssh public keys for the default user (defined in ``users``)) is
```
Comment (OddBloke) [inline cloudinit/config/cc_ssh.py]: I think this paragraph you've inherited from the existing docs is just plain wrong. It's the only mention of `allow_publish_ssh_keys` in the entire project.

File: cloudinit/config/cc_ssh.py (hunk 7 above)
Comment (OddBloke) [reply cloudinit/config/cc_ssh.py]: This expanded sentence will make it a little easier for people to key in on this being the thing they are looking for, I think:

```suggestion
connect to a user account on a system. They are stored in `.ssh/authorized_keys` in that account's home directory. Authorized keys for the default user/first user defined
```

Comment (powersj) [reply cloudinit/config/cc_ssh.py]: Done - I do think we are now in the realm of re-writting this module's docs :)

File: cloudinit/config/cc_ssh.py (hunk 8 above)
Comment (powersj) [reply cloudinit/config/cc_ssh.py]: Agreed, I think I would normally say we could follow up, but I've removed it for now.

**Comment by powersj** [comment]
Ready for another review

**Pull request merged**